File: src/zing.py
#!/usr/bin/env python3
#
#

"""

 Name:  zing.py
 
 Title: Zing - Zero packet pING network utility implemented in Python.

 Description: Zero packet PING utility that checks host by name or ip-address is active, and time to reach.

 Author William F. Gilreath (will@wfgilreath.xyz)
 Version 1.2.2  10/12/22

 Copyright © 2022 All Rights Reserved.

 License: This software is subject to the terms of the GNU General Public License (GPL)
     version 3.0 available at the following link: http://www.gnu.org/copyleft/gpl.html.

 You must accept the terms of the GNU General Public License (GPL) license agreement
     to use this software.

 Note Python version of zing, the first, initial time value can skew overall results, so the first value
     in zing timetable is skipped for computing average, standard deviation, extremum.

"""

import math
import socket
import sys
import time

class Zing:

    ports = [80, 443]   # default ports http, https
    inet_addr = None    # network address name for hostname
    tcp4Flag = True     # default tcp4 ip-address
    timeout = 4000      # default socket time 4000 ms = 4-seconds
    host = "localhost"  # default host name is localhost or 127.0.0.1
    count = 4           # default count of times to perform ops
    host_name = None    # result host name from DNS query
    host_addr = None    # result host address from DNS query
    limit = 4           # default limit on number of ops

    #
    # Get total time to zing using equation: time = (double) timeTotal / (double) ports.length / (double) limit;
    #
    # parameter: in_total_time is the accumulated time overall.
    # parameter: in_ports_len is the total number of ports.
    # parameter: in_limit is the limit on the number of ops per zing request.
    # 
    # return - total time to zing host averaged with equation.
    #
    @staticmethod
    def get_total_time(in_total_time, in_ports_len, in_limit):
        result = in_total_time / in_ports_len / in_limit
        return result

    #
    # Get the wallclock real-time in milliseconds
    #
    # return - current wallclock time in milliseconds
    #
    @staticmethod
    def current_time_ms():
        return time.clock_gettime_ns(time.CLOCK_REALTIME) / 1000

    #
    # Calculate the standard deviation using the average time and the table of
    # zing times. Standard deviation is the statistical measure of variability.
    #
    # parameter: avg - average time to zing a computer system on the network.
    # parameter: values - the zing table of times for a count of each time to zing a computer system.
    #
    # return: double - computed standard deviation or variance in values
    #
    @staticmethod
    def stddev(in_avg, in_values):
        dv = 0.0
        for dbl in enumerate(in_values):
            if dbl[0] == 0:  # skip first value as skews standard deviation calculation
                continue
            dm = dbl[1] - in_avg
            dv = dv + (dm * dm)
        result = math.sqrt(dv / len(in_values) - 1)
        result = round(result, 3)
        return result

    #
    # Report or print the use of zing with parameters and example with default parameters
    #
    def usage(self):
        print("Usage: zing.py -h | [-4|-6] [-c count] [-op ops] [-p ports] [-t timeout] host")
        print("zing.py -4 -c 4 -op 4 -p 80,443 -t 4000 google.com")

    #
    # Process command-line arguments and set internal zing parameters from values.
    #
    # parameter: argv - command-line interface arguments passed to zing network utility
    #
    def process_args(self, argv):
        n = len(sys.argv)
        idx = 1
        while idx < n:
            arg = sys.argv[idx]

            if arg == "-4":
                self.tcp4Flag = True
            elif arg == "-6":
                self.tcp4Flag = False
            elif arg == "-c":
                self.count = sys.argv[idx + 1]
                idx = idx + 1
            elif arg == "-h":
                self.usage()
                exit(0)
            elif arg == "-op":
                self.limit = sys.argv[idx + 1]
                idx = idx + 1
            elif arg == "-p":
                self.ports = sys.argv[idx + 1].split(",")
                idx = idx + 1
            elif arg == "-t":
                self.timeout = sys.argv[idx + 1]
                idx = idx + 1
            else:
                if arg[0] == '-':
                    print("Error! Invalid command-line argument:", arg, "not recognized!")
                    exit(1)
                else:
                    self.host = arg
            idx = idx + 1

    #
    # Get host IP-address and host name
    #
    # parameter: in_host - specified host to zing
    #
    # return IP address as IPv4 or IPv6
    #
    def get_host_addr_name(self, in_host):
        try:
            if self.tcp4Flag:
                return self.get_host_addr_name_ip4(in_host)
            else:
                return self.get_host_addr_name_ip6(in_host)
        except Exception:
            print("Error! Unable to get host'", in_host, end="' ")
            print("host address or ip-address unknown!")
            exit(1)

    #
    # Get host IP-address and host name as IPv4
    #
    # parameter: in_host - specified host to zing
    #
    # return IP address as IPv4
    #
    def get_host_addr_name_ip4(self, in_host):
        ip_addr = socket.gethostbyname(in_host)
        self.host_addr = ip_addr
        self.host_name = in_host
        return ip_addr

    #
    # Get host IP-address and host name as IPv6
    #
    # parameter: in_host - specified host to zing
    #
    # return IP address as IPv6
    #
    def get_host_addr_name_ip6(self, in_host):
        ip_addr = self.get_ip_6(in_host)
        self.host_addr = ip_addr
        self.host_name = in_host
        return ip_addr

    #
    # Get host IP-address v6
    #
    # parameter: in_host - specified host to zing
    # parameter: in_port - specified port to use with default port as 0
    #
    # return IP address as IPv6
    #
    @staticmethod
    def get_ip_6(in_host, in_port=0):
        result = socket.getaddrinfo(in_host, in_port, socket.AF_INET6)
        ip_addr = result[0][4][0]
        return ip_addr

    #
    # Get host IP-address v4
    #
    # parameter: in_host - specified host to zing
    # parameter: in_port - specified port to use with default port as 0
    #
    # return IP address as IPv4
    #
    @staticmethod
    def get_ip_4(in_host, port=0):
        result = socket.getaddrinfo(in_host, port, socket.AF_INET)
        ip_addr = result[0][4][0]
        return ip_addr

    #
    # Report the time and if the host computer system is active or alive on the network.
    #
    # parameter: in_time - overall time to zing the host computer system on the network.
    #
    def report_time(self, in_time):
        print("", end=" ")
        print(int(self.limit) * len(self.ports), end=" ")
        print("ops to", end=" ")
        print(self.host_name, end=" ")
        print("(", end="")
        print(self.host_addr, end="")
        print(")", end=" ")

        if in_time >= 0.0:
            print("Active", end=" ")
            print("time", end=" = ")
            print(round(in_time, 3), end=" ")
            print("ms ")
        else:
            print("Absent", end="!")
            exit(0)

    #
    # Zing a given host on the network at a specific port on the host.
    #
    # parameter: in_host - host name of computer system on a network.
    # parameter: in_port - port on the computer system on a network.
    #
    # return: double - total socket time to zing computer system or -1.0d for  not available.
    #
    def do_zing_to_host(self, in_host, in_port):
        if self.inet_addr is None:
            self.inet_addr = self.get_host_addr_name(in_host)

        self.host_addr = self.inet_addr
        self.host_name = in_host

        present_flag = True

        sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM, 0)
        sock.settimeout(int(self.timeout))

        # get start time in milliseconds
        sock_time_start = self.current_time_ms()

        # sock.connect((self.inet_addr, int(self.ports[0]), 0, 0))
        sock.connect((self.inet_addr, int(in_port), 0, 0))

        sock.shutdown(socket.SHUT_RDWR)
        sock.close()

        # get close time in milliseconds
        sock_time_close = self.current_time_ms()

        if present_flag:
            sock_time_total = sock_time_close - sock_time_start
        else:
            print("", end=".")
            return -1.0

        return sock_time_total

    #
    # The main method is the central start method of the zing network utility
    # that invokes other methods to zing a host computer system on a network.
    #
    @staticmethod
    def main():
        zing = Zing()
        n = len(sys.argv)

        if n == 1:
            zing.usage()
            exit(0)

        zing.host = sys.argv[1]
        zing.process_args(sys.argv)

        zing.get_host_addr_name(zing.host)

        print("ZING:", zing.host_name, end=" (")
        print(zing.host_addr, end="):")
        print(" ", len(zing.ports), end=" ")
        print("ports used,", end=" ")
        print((int(zing.limit) * len(zing.ports)), end=" ")
        print("ops per cycle")

        zing_time_table = []
        time_zing_start = zing.current_time_ms()

        for x in range(int(zing.count) + 1):  # add one to count as we skip first entry hence need 1...count+1
            total_time = 0

            # skip reporting/printing timing for first zero-th 0 cycle of the for-loop
            if x > 0:
               print("#", x, end=".")

            for y in range(int(zing.limit)):
                for port in enumerate(zing.ports):
                    zing_time = zing.do_zing_to_host(zing.host, int(port[1]))
                    zing_time_table.append(zing_time)
                    total_time = total_time + zing_time

            if x > 0:
                print(".", end="")

            total_time = zing.get_total_time(int(total_time), len(zing.ports), int(zing.limit))

            if x > 0:
                print(".", end="")

            if x > 0:
                zing.report_time(total_time)

        time_zing_close = zing.current_time_ms()

        min_time = int(zing.timeout)
        max_time = 0
        avg_time = 0

        for val in enumerate(zing_time_table):
            # skip first value as initial time is large and skews results
            if val[0] == 0:  
                continue
            if val[1] > max_time:
                max_time = val[1]

            if val[1] < min_time:
                min_time = val[1]

            avg_time = avg_time + val[1]

        avg_time = avg_time / (len(zing_time_table) - 1)
        std_dev = zing.stddev(avg_time, zing_time_table)

        print()
        print("--- zing summary for", end=" ")
        print(zing.host_name, end="/")
        print(zing.host_addr, end=" ---")
        print()
        print(len(zing.ports) * int(zing.limit), end=" ")
        print("total ops used; total time", end=": ")
        print(time_zing_close - time_zing_start, end=" ms")
        print()
        print("total-time min/avg_time/max/stddev ", end="= ")
        print(round(min_time, 3), end="/")  
        print(round(avg_time, 3), end="/")
        print(round(max_time, 3), end="/")
        print(std_dev, end=" ms")
        print()
        print()


if __name__ == '__main__':
    try:
        sys.exit(Zing.main())
    except Exception as ex:
        import traceback

        exc_type, exc_val, exc_trace = sys.exc_info()
        print("Error! ", exc_val)
        print(exc_type, exc_val)
        print("Trace:")
        traceback.print_tb(exc_trace, file=sys.stdout)
        sys.exit(1)

File: src/Zing.cs
// ==++==
// 
//   Copyright (c) 2022 by William F. Gilreath (will@wfgilreath.xyz).
//   All rights reserved.
//
//   License: This software is subject to the terms of the GNU General Public License (GPL)  
//   version 3.0 available at the following link: http://www.gnu.org/copyleft/gpl.html.
//
//   You must accept the terms of the GNU General Public License (GPL) license agreement
//   to use this software.
//
// ==--==
/*============================================================
**
** Class: Zing - Zero packet pING network utility.
**
**
** Purpose: Zero packet PING utility that checks host by name or ip-address 
**          is active, and time in milliseconds to reach.
**
** Version: 1.2.2 November 2022
**
**
============================================================*/
namespace xyz.wfgilreath.net
{
    using System.Net;
    using System.Net.Sockets;

    public sealed class Zing : Object
    {

        private static IPAddress? inet_addr = null; // network address name for hostname

        private static bool tcp4Flag = true; // default tcp4 ip-address
        private static int  timeout  = 4000; // default socket time 4000 ms = 4-seconds
        private static int  count    = 4;    // default count of times to perform ops

        private static int[] ports   = new int[] { 80, 443 }; // default ports http, https

        private static String  host     = "localhost";   // default host name is localhost or 127.0.0.1
        private static String? hostName = String.Empty;  // result host name from DNS query
        private static String? hostAddr = String.Empty;  // result host address from DNS query
        private static bool    hostFlag = true;          // default is host is present, available
        private static int     limit    = 4;             // default limit on number of ops

        // zing usage and example parameters from defaults defined in code
        private const String   ZING_USAGE   = "Usage: zing /h | [/4|/6] [/c count] [/op ops] [/p ports] [/t timeout] host";
        private const String ZING_EXAMPLE   = "zing /4 /c 4 /op 4 /p 80,443 /t 4000 microsoft.com";

        private const String TIME_FORMAT    = "{0:0.000}"; //time format for reporting zing host time

        private static void dumpArgs()
        {
            Console.Write("tcp4Flag: "); Console.WriteLine(tcp4Flag.ToString());
            Console.Write("timeout:  "); Console.WriteLine( timeout.ToString());
            Console.Write("count:    "); Console.WriteLine(   count.ToString());
            Console.Write("host:     "); Console.WriteLine(    host.ToString());
            Console.Write("limit:    "); Console.WriteLine(   limit.ToString());

            Console.Write("ports:    ");

            foreach (int port in ports)
            {
                Console.Write(port);
                Console.Write(" ");
            }//end foreach

            Console.WriteLine();

        }//end dumpArgs

        private static void Usage()
        {
            Console.WriteLine();
            Console.WriteLine(ZING_USAGE);
            Console.WriteLine();
            Console.WriteLine(ZING_EXAMPLE);
            Console.WriteLine();
            Console.WriteLine();

        }//end usage

        private static void ProcessArgs(String[] args)
        {
            try
            {
                for (int idx = 0; idx < args.Length; idx++)
                {
                    String arg = args[idx];
                    switch (arg)
                    {

                        case "-4":
                        case "/4":
                            tcp4Flag = true;
                            break;

                        case "-6":
                        case "/6":
                            tcp4Flag = false;
                            break;

                        case "-c":
                        case "/c":
                            count = int.Parse(args[idx + 1]);
                            idx++;
                            break;

                        case "-op":
                        case "/op":
                            limit = int.Parse(args[idx + 1]);
                            idx++;
                            break;

                        case "-p":
                        case "/p":

                            String[] tmpPorts = args[idx + 1].Split(","); // split comma delimited list

                            ports = new int[tmpPorts.Length];
                            for (int x = 0; x < tmpPorts.Length; x++)
                            {
                                ports[x] = int.Parse(tmpPorts[x]); 
                            } // end for

                            break;


                        case "-t":
                        case "/t":
                            timeout = int.Parse(args[idx + 1]);
                            idx++;
                            break;

                        case "-h":
                        case "/h":
                            Usage();
                            Environment.Exit(0);
                            break;

                        // check if arg has "/" or "-" at char[0], if so error invalid command-line parameter
                        default:

                            char prefixChar = arg.ToCharArray()[0];

                            if(prefixChar == '/' || prefixChar == '-')
                            {
                                Console.Write("Error ");
                                Console.Write(arg);
                                Console.WriteLine(" is invalid command-line parameter!");
                                Environment.Exit(1);
                            }
                            else
                            {
                                host = arg;
                            } // end if

                            break;

                    }// end switch

                }//end for
            }
            catch (Exception)
            {
                Console.WriteLine("Error with command-line arguments!%n");
                Environment.Exit(1);

            }//end try

        }//end processArgs

        public static IPAddress? GetIPv4Addr(String hostname)
        {

            // Get server related information.
            IPHostEntry heserver = Dns.GetHostEntry(hostname);

            // Loop on the AddressList
            foreach (IPAddress addr in heserver.AddressList)
            {

                if(addr.AddressFamily.ToString() == ProtocolFamily.InterNetwork.ToString())
                {
                    hostName = hostname;
                    return addr;
                }

            }//end foreach

            hostName = hostname;
            return null;

        }//end getIPv4Addr

        public static IPAddress? GetIPv6Addr(String hostname)
        {
            // Get server related information.
            IPHostEntry heserver = Dns.GetHostEntry(hostname);

            // Loop on the AddressList
            foreach (IPAddress addr in heserver.AddressList)
            {

                if (addr.AddressFamily.ToString() == ProtocolFamily.InterNetworkV6.ToString())
                {
                    //hostName = heserver.HostName;
                    hostName = hostname; 
                    return addr;
                }

            }//end foreach

            hostName = hostname;
            return null;

        }//end getIPv6Addr

        public static IPAddress? GetHostAddrName(String host) //use struct to return multiple values
        {

            IPAddress? iaddr = null;
            try
            {

                if(tcp4Flag)
                {
                    if (Socket.OSSupportsIPv4)
                    {
                        iaddr = GetIPv4Addr(host);
                    }
                    else
                    {
                        Console.WriteLine("OS does not support IPv4!"); 
                        return null;
                    }//end if
                }
                else
                {
                    if (Socket.OSSupportsIPv6)
                    {
                        iaddr = GetIPv6Addr(host);
                    }
                    else
                    {
                        Console.WriteLine("OS does not support IPv6!");
                        return null;
                    }//end if
                } // end if

                if (hostFlag)
                {
                    if (iaddr != null)
                    {
                       
                        hostAddr = iaddr.ToString();
                    }
                    else
                    {
                        throw new Exception(".. Error: Cannot resolve " + Zing.host + " Unknown host.");
                    }//end if

                    hostFlag = false;

                } // end if
            }
            catch (Exception)
            {
                Console.Write(".. Error: Cannot resolve ");
                Console.Write(Zing.host);
                Console.WriteLine(" Unknown host.");
                Environment.Exit(1);
            } // end try

            return iaddr;

        }// end getHostAddrName

        public static void Report(in double time)
        {
            Console.Write(" ");
            Console.Write(limit*ports.Length);
            Console.Write(" ops to ");
            Console.Write(hostName);
            Console.Write(" (");
            Console.Write(hostAddr);
            Console.Write("); ");

            if (time >= 0.0d)
            {
                Console.Write("Active ");
                Console.Write("time = ");
                //Console.Write(time);
                Console.Write(String.Format("{0:0.000}", time));
                Console.WriteLine(" ms");
            }
            else
            {
                Console.Write("Absent!");
                Environment.Exit(0);
            } // end if

        }// end report

        public static double GetTotalTime(in double totalTime, in int portsLength, in int limit)
        {
            double time = totalTime / (double)portsLength / (double)limit;
            return time;
        }// end getTotalTime

        public static double DoZingToHost(in String host, in int port)
        {

            if (inet_addr == null)
            {
                inet_addr = GetHostAddrName(host);
            } // end if

            bool presentFlag = true; // host at socket is present, default is true

            long socketTimeStart = 0, socketTimeClose = 0, socketTimeTotal = 0;

            try
            {
                if(inet_addr == null)
                {
                    Console.WriteLine("Error: Unable to get IP address!");
                    return 0.0d;
                }

                IPEndPoint ip_ep = new(inet_addr, port);

                socketTimeStart  = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;

                Socket sock      = new(inet_addr.AddressFamily,
                                       SocketType.Stream,
                                       ProtocolType.Tcp){
                    SendTimeout  = Zing.timeout
                };

                sock.Connect(ip_ep);

                sock.Close();

                socketTimeClose = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;

            }
            catch (SocketException se)
            {
                Console.WriteLine("Socket Error : {0}", se.ToString());
                presentFlag = false;
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error : {0}", ex.ToString());
                presentFlag = false;
            }//end try

            if (presentFlag)
            {
                socketTimeTotal = socketTimeTotal + (socketTimeClose - socketTimeStart);
            }
            else
            {
                Console.Write(".");
                return -1.0d;
            } // end if

            return (double)socketTimeTotal;

        }//end doZingToHost


        public static double Stddev(in double avg, double[] values)
        {
            double dv = 0;

            foreach (double dbl in values)
            {
                double dm = dbl - avg;
                dv += dm * dm;
            }//end for
            
            return Math.Sqrt(dv / values.Length);

        }//end stddev

        public static void Main(string[] args)
        {

            if (args.Length == 0)
            {
                Usage();
                Environment.Exit(0);

            }//end if

            host = args[0];

            ProcessArgs(args);

            Zing.inet_addr = GetHostAddrName(host);

            Console.Write("ZING: ");
            Console.Write(hostName);
            Console.Write(" (");
            Console.Write(hostAddr);
            Console.Write("): ");
            Console.Write(ports.Length);
            Console.Write(" ports used, ");
            Console.Write((limit * ports.Length));
            Console.Write(" ops per cycle");
            Console.WriteLine();

            long timeZingStart = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;

            Double[] zingTimeTable = new double[count];

            for (int x = 0; x < count; x++)
            {
                double totalTime = 0.0;

                Console.Write("#");
                Console.Write(x + 1);
                Console.Write(" ");

                Console.Write(".");

                for (int y = 0; y < limit; y++)
                {
                    foreach(int port in ports)
                    {
                        double zingTime = DoZingToHost(host, port);
                        zingTimeTable[x] = zingTime;
                        totalTime += zingTime;
                    } // end for(port)
                } // end for(limit)

                Console.Write(".");

                double time = GetTotalTime(totalTime, ports.Length, limit);

                Console.Write(".");

                Report(time); // time = -1.0d, absent, else active

            } // end for(count)

            long timeZingClose = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;

            //get min, max, avg
            double min = Double.MaxValue, max = Double.MinValue, avg = 0.0d;
            for (int x = 0; x < count; x++)
            {

                if (min > zingTimeTable[x])
                {
                    min = zingTimeTable[x];
                }//end if

                if (max < zingTimeTable[x])
                {
                    max = zingTimeTable[x];
                }//end if

                avg += zingTimeTable[x];

            }//end for

            avg /= (double) count;

            double std_dev = Stddev(avg, zingTimeTable);

            Console.WriteLine();
            Console.Write("--- zing summary for ");
            Console.Write(hostName);
            Console.Write("/");
            Console.Write(hostAddr);
            Console.Write("---");
            Console.WriteLine();

            Console.Write(ports.Length * limit * count);
            Console.Write(" total ops used; total time: ");
            Console.Write(timeZingClose - timeZingStart);
            Console.WriteLine(" ms");

            Console.Write("total-time min/avg/max/stddev = ");
            Console.Write(String.Format(TIME_FORMAT, min));
            Console.Write("/");
            Console.Write(String.Format(TIME_FORMAT, avg));
            Console.Write("/");
            Console.Write(String.Format(TIME_FORMAT, max));
            Console.Write("/");
            Console.Write(String.Format(TIME_FORMAT, std_dev));
            Console.WriteLine(" ms");
            Console.WriteLine();
            Console.WriteLine();

            Environment.Exit(0);

        }//end Main

    }//end class Zing

}//end namespace